WS2812.class.nut

```diff
@@ -2,121 +2,125 @@
-	{
-		_spi.write(_frame);
-	}
-	
-	// ------ PRIVATE METHODS - DO NOT CALL DIRECTLY ------
-	
-	// Fill the array of representative 1-wire waveforms,
-	// done by the constructor at instantiation
-	
-	function _initialize() 
-	{
-		// Fill the bits array first
-
-		_bits = array(256);
-		
-		for (local i = 0; i < 256; i++) 
-		{
-			local valblob = blob(BYTESPERPIXEL / 3);
-			valblob.writen((i & 0x80) ? ONE:ZERO,'b');
-			valblob.writen((i & 0x40) ? ONE:ZERO,'b');
-			valblob.writen((i & 0x20) ? ONE:ZERO,'b');
-			valblob.writen((i & 0x10) ? ONE:ZERO,'b');
-			valblob.writen((i & 0x08) ? ONE:ZERO,'b');
-			valblob.writen((i & 0x04) ? ONE:ZERO,'b');
-			valblob.writen((i & 0x02) ? ONE:ZERO,'b');
-			valblob.writen((i & 0x01) ? ONE:ZERO,'b');
-			_bits[i] = valblob;
-		}
-
-		// Now fill the clearblob
-		
-		_clearblob = blob(BYTESPERPIXEL);
-		
-		for (local j = 0 ; j < BYTESPERPIXEL ; j++) 
-		{
-			_clearblob.writen(ZERO, 'b');
-		}
-	}
+class WS2812 {
+    // This class uses SPI to emulate the WS2812s' one-wire protocol.
+    // This requires one byte per bit to send data at 7.5 MHz via SPI.
+    // These consts define the "waveform" to represent a zero or one
+
+    static version = [2,0,0];
+
+    static ZERO            = 0xC0;
+    static ONE             = 0xF8;
+    static BYTES_PER_PIXEL   = 24;
+
+    // When instantiated, the WS2812 class will fill this array with blobs to
+    // represent the waveforms to send the numbers 0 to 255. This allows the blobs to be
+    // copied in directly, instead of being built for each pixel - which makes the class faster.
+
+    _bits            = null;
+
+    // Private variables passed into the constructor
+
+    _spi             = null;  // imp SPI interface (pre-configured)
+    _frameSize       = null;  // number of pixels per frame
+    _frame           = null;  // a blob to hold the current frame
+
+    // Parameters:
+    //    spi          A pre-configured SPI bus (MSB_FIRST, 7500)
+    //    frameSize    Number of Pixels per frame
+
+    constructor(spiBus, frameSize)
+    {
+        // spiBus must be configured
+        _spi = spiBus;
+
+        _frameSize = frameSize;
+        _frame = blob(_frameSize * BYTES_PER_PIXEL + 1);
+        _frame[_frameSize * BYTES_PER_PIXEL] = 0;
+
+        // Prepare the bits array and the clearblob blob
+
+        _initialize();
+
+        // Zero the LED array
+
+        fill([0,0,0]);
+        draw();
+    }
+
+    // ------- PUBLIC FUNCTIONS -------
+
+    // Sets a pixel in the buffer
+    //   p - the pixel (0 <= p < _frameSize)
+    //   color - [r,g,b] (0 <= r,g,b <= 255)
+    function set(p, color) {
+        assert(p >= 0);
+        assert(p < _frameSize);
+
+        _frame.seek(p * BYTES_PER_PIXEL);
+
+        // Red and green are swapped for some reason, so swizzle them back
+        _frame.writeblob(_bits[color[1]]);
+        _frame.writeblob(_bits[color[0]]);
+        _frame.writeblob(_bits[color[2]]);
+    }
+
+
+    // Clears the frame buffer
```

Comment:
Comments need updating since this is no longer "clear". Probably want a comment which says fill can be used for a clear.